lib/core/src/connection_pool.cpp

```diff
@@ -87,23 +88,61 @@ namespace irods
                                      const std::string& _username,
                                      const std::string& _zone,
                                      const int _refresh_time)
+        : connection_pool{_size, _host, _port, "", "", _username, _zone, _refresh_time, {}}
+    {
+    }
+
+    connection_pool::connection_pool(int _size,
+                                     const std::string& _host,
+                                     const int _port,
+                                     const std::string& _username,
+                                     const std::string& _zone,
+                                     const int _refresh_time,
+                                     std::function<void(RcComm&)> _auth_func)
+        : connection_pool{_size, _host, _port, "", "", _username, _zone, _refresh_time, std::move(_auth_func)}
+    {
+    }
+
+    connection_pool::connection_pool(int _size,
+                                     std::string_view _host,
+                                     const int _port,
+                                     std::string_view _proxy_username,
+                                     std::string_view _proxy_zone,
+                                     std::string_view _username,
+                                     std::string_view _zone,
+                                     const int _refresh_time,
+                                     std::function<void(RcComm&)> _auth_func)
         : host_{_host}
         , port_{_port}
+        , proxy_username_{_proxy_username}
+        , proxy_zone_{_proxy_zone}
         , username_{_username}
         , zone_{_zone}
         , refresh_time_(_refresh_time)
+        , auth_func_{std::move(_auth_func)}
         , conn_ctxs_(_size)
     {
         if (_size < 1) {
             throw std::runtime_error{"invalid connection pool size"};
         }
 
+        if (!proxy_username_.empty() && proxy_zone_.empty()) {
+            throw std::runtime_error{"invalid proxy zone"};
+        }
+
+        if (proxy_username_.empty() && !proxy_zone_.empty()) {
+            throw std::runtime_error{"invalid proxy username"};
+        }
+
+        const auto use_proxy_connection = (!proxy_username_.empty() && !proxy_zone_.empty());
```

Comment:
Consider designing a better approach than relying on the emptiness of these strings.

Reply:
Probably better to make these `std::optional<std::string>`s instead of strings. It better expresses the intent and leaves room to treat an empty string as a valid argument.

May need to introduce a `user` type  (and a `proxy_user` type) to hold this info. However, we already have https://github.com/irods/irods/blob/1205fa3afd8de1e96e0e412184c7b30ae0b07c31/lib/administration/user/include/irods/user.hpp.

Reply:
Implemented `unique_user` class to enforce constraints for user identity info and connecting to a server.